File: InsertionSort/InsertionSort.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

void insertionSort(int *toSort, int size);
static unsigned long getNanos(void);

int main(void)
{
  time_t t1,t2;
  int elapsedTime;
  int arraySize = 100000;
  int toSort[arraySize];
  srand((unsigned int)arraySize);
  unsigned long start,end;

  for(int i = 0; i < 20; i++)
    {
      for(int j = 0; j < arraySize; j++)
	{
	  toSort[j] = rand() % (arraySize * 10);
	}

      start = getNanos();
      insertionSort(toSort, arraySize);
      end = getNanos();

      fprintf(stdout, "Sorting took %ld\n", (end-start));

    }
  return 1;
}

void insertionSort(int *toSort, int size)
{
  //Loop over the array (the first element is sorted
  for(int i =1; i < size; i++)
    {
      int move = i;
      while((move > 0) && (toSort[move-1] > toSort[move]))
	{
	  toSort[move-1] += toSort[move];
	  toSort[move] = toSort[move-1] - toSort[move];
	  toSort[move-1] = toSort[move-1] - toSort[move];
	  move--;
	}
    }
}

static unsigned long getNanos(void) {
  struct timespec ts;
  timespec_get(&ts, TIME_UTC);
  return (unsigned long)ts.tv_sec * 1000000000L + ts.tv_nsec;
}  

File: InsertionSort/InsertionSort.java
import java.util.Random;

class InsertionSort
{
    public static void main(String[] args)
    {
	int arraySize = 100000;
	int[] toSort = new int[arraySize];
	Random rand = new Random(arraySize);
	for(int j = 0; j < 20; j++)
	    {
		for(int i = 0; i < toSort.length; i++)
		    {
			toSort[i] = rand.nextInt(arraySize * 10);
		    }
		Long start = System.nanoTime();
		InsertionSort.sort(toSort);
		Long end = System.nanoTime();
		System.out.println("Sorting took: " + (end - start));
		//for(int k = 0; k < 100; k++)
		//    {
		//	System.out.print(toSort[k]);
		//    }
		//System.out.println();
	    }
    }

    private static int[] sort(int[] toSort)
    {
	//Loop over all elements (first one not needed)
	for(int i = 1; i < toSort.length; i++)
	    {
		int move = i;
		while((move > 0) && (toSort[move-1] > toSort[move]))
		    {
			toSort[move-1] += toSort[move];
			toSort[move] = toSort[move-1] - toSort[move];
			toSort[move-1] = toSort[move-1] - toSort[move];
			move--;
		    }
	    }
	return toSort;
    }
}

